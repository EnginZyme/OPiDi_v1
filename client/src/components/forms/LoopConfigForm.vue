<template>
  <div class="step_config_form" v-if="formType === 'loop'">
    <p class="step_config_form__title">Loop</p>
    <div class="step_config_form__body">
      <div class="step_config_form__body__content">
        <div>
          <p class="step_config_form__body__content__label--medium">
            Step Name
          </p>
          <InputText
            type="text"
            class="step_config_form__body__content__input--long"
            v-model="loopStepName"
          />
        </div>
        <div>
          <p class="step_config_form__body__content__label--medium">
            Number of Loops
          </p>
          <InputNumber
            :min="1"
            class="step_config_form__body__content__input--short"
            v-model="loopNumber"
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import types from "../../types";
import { mapGetters } from "vuex";
import { findDeep } from "../../utils";

/**
 * Accepts user input for the configuration of a loop step within the object in the protocol cache (editor).
 */
export default {
  name: "LoopConfigForm",
  methods: {
    /**
     * Fetches a protocol step by its ID.
     * 
     * @params {number} id - The said ID.
     */
    getProtocolStep(id) {
      return findDeep(this.cachedProtocolSteps, id);
    },
  },
  props: {
    /**
     * Determines if this component will be rendered or not (true if its value is set to `loop`).
     */
    formType: String,
    /**
     * The ID of the current user-selected step card.
     */
    stepID: Number,
  },
  computed: {
    ...mapGetters(["cachedProtocolSteps", "nonTipRackLabwares"]),
    /**
     * A read/write object for the name of the user-selected loop step.
     */
    loopStepName: {
      get: function() {
        return this.formType !== "loop" || this.stepID < 0
          ? null
          : this.getProtocolStep(this.stepID).name;
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_LOOP_STEP_NAME, {
          id: this.stepID,
          newValue,
        });
      },
    },
    /**
     * A read/write object for the number of loops of the user-selected loop step.
     */
    loopNumber: {
      get: function() {
        return this.formType !== "loop" || this.stepID < 0
          ? null
          : this.getProtocolStep(this.stepID).num_iterations;
      },
      set: function(newValue) {
        this.$store.dispatch(types.UPDATE_LOOP_NUMBER, {
          id: this.stepID,
          newValue: parseInt(newValue),
        });
      },
    },
  },
};
</script>

<style scoped>
.step_config_form {
  display: grid;
  grid-template-rows: auto auto;
  padding-top: 20px;
  padding-bottom: 10px;
}
.step_config_form__body {
  border-radius: 5px;
  box-shadow: 0 -1em 1em -1em #bcbcbc inset;
  padding-top: 20px;
  padding-left: 16px;
  display: grid;
  overflow: auto;
  height: 65vh;
  margin-top: 10px;
}
.step_config_form__body__content {
  align-self: center;
  display: grid;
  grid-row-gap: 3vh;
}
.step_config_form__body__content__input--short {
  background: #efefff;
  box-shadow: inset 2px 2px 4px rgba(0, 0, 0, 0.25),
    inset -2px -2px 4px 3px rgba(255, 255, 255, 0.25),
    inset -2px -2px 4px 3px rgba(255, 255, 255, 0.25);
  border-radius: 6px;
  font-size: 12px;
  width: 40%;
  height: 32px;
  border: none;
}
.step_config_form__body__content__input--long {
  background: #efefff;
  box-shadow: inset 2px 2px 4px rgba(0, 0, 0, 0.25),
    inset -2px -2px 4px 3px rgba(255, 255, 255, 0.25),
    inset -2px -2px 4px 3px rgba(255, 255, 255, 0.25);
  border-radius: 6px;
  font-size: 12px;
  width: 91%;
  height: 32px;
  border: none;
}
.step_config_form__title {
  font-size: 18px;
  margin: 0px;
}
.step_config_form__body__content__label--medium {
  font-size: 13px;
  margin: 0px;
  margin-bottom: 8px;
}
</style>
