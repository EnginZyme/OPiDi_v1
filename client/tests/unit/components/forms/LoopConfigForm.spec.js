import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import LoopConfigForm from "@/components/forms/LoopConfigForm";

const localVue = createLocalVue();

localVue.use(Vuex);

const sampleFormType = "loop";
const sampleStepID = 1;
const sampleCachedProtocolSteps = [
  {
    type: "loop",
    name: "Loop (1)",
    id: 1,
    num_iterations: 3,
    substeps: [
      {
        type: "simple_transfer",
        name: "Simple Transfer (2)",
        id: 2,
        substeps: [],
        parameters: {
          pipette: "right_pipette",
          pipette_obj: {
            name: "(Right) Single Channel 300 uL - Gen 1",
            category: "right_pipette",
          },
          pipette_strategy: "",
          volumes: [100],
          volumes_string: "100",
          source: { slot_number: 5, slot_name: "reservoir_5", wells: ["A1"] },
          destination: {
            slot_number: 1,
            slot_name: "wellplate_1",
            wells: ["D6"],
          },
          liquidClass: "Default",
          liquidClassId: 0,
          tipsStrategy: "ECO",
          offset: {
            source: 1,
            destination: 1,
            source_type: "Bottom",
            destination_type: "Bottom",
          },
        },
      },
    ],
  },
];
const sampleNonTipRackLabwares = [
  {
    name: "wellplate_1",
    labware_name: "test 96 wellplate 2ml 3xT",
    labware_type: "shakerLabware",
    labware_id: "test_96_wellplate_2ml_3xT",
    nickname: "",
    display_name: "test 96 wellplate 2ml 3xT [1]",
    id: 1,
  },
  {
    name: "tuberack_2",
    labware_name: "test tuberack 65 mL tubes",
    labware_type: "tubeRack",
    labware_id: "test_tuberack_65_mL_tubes",
    nickname: "",
    display_name: "test tuberack 65 mL tubes [2]",
    id: 2,
  },
  {
    name: "reservoir_5",
    labware_name: "test slot 3 mL",
    labware_type: "reservoir",
    labware_id: "test_slot_3_mL",
    nickname: "",
    display_name: "test slot 3 mL [5]",
    id: 5,
  },
];

describe("shallowMounted LoopConfigForm", () => {
  let getters;
  let store;
  let wrapper;

  beforeEach(() => {
    getters = {
      cachedProtocolSteps: () => sampleCachedProtocolSteps,
      nonTipRackLabwares: () => sampleNonTipRackLabwares,
    };

    store = new Vuex.Store({
      getters,
    });

    wrapper = shallowMount(LoopConfigForm, {
      store,
      localVue,
      propsData: {
        formType: sampleFormType,
        stepID: sampleStepID,
      },
    });
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
