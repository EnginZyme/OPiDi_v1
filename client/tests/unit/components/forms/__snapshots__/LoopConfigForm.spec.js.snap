// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallowMounted LoopConfigForm renders correctly 1`] = `
<div
  class="step_config_form"
>
  <p
    class="step_config_form__title"
  >
    Loop
  </p>
   
  <div
    class="step_config_form__body"
  >
    <div
      class="step_config_form__body__content"
    >
      <div>
        <p
          class="step_config_form__body__content__label--medium"
        >
          
          Step Name
        
        </p>
         
        <inputtext
          class="step_config_form__body__content__input--long"
          type="text"
        />
      </div>
       
      <div>
        <p
          class="step_config_form__body__content__label--medium"
        >
          
          Number of Loops
        
        </p>
         
        <inputnumber
          class="step_config_form__body__content__input--short"
          min="1"
        />
      </div>
    </div>
  </div>
</div>
`;
